src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Centipede engine interface."""
+
+import glob
+import os
+from pathlib import Path
+import re
+
+from clusterfuzz._internal.bot.fuzzers import dictionary_manager
+from clusterfuzz._internal.bot.fuzzers import engine_common
+from clusterfuzz._internal.bot.fuzzers import utils as fuzzer_utils
+from clusterfuzz._internal.system import environment
+from clusterfuzz._internal.system import new_process
+from clusterfuzz.fuzz import engine
+
+_CLEAN_EXIT_SECS = 10
+_TIMEOUT = 1200
+_SERVER_COUNT = 1
+_RSS_LIMIT = 4096
+_RLIMIT_AS = 5120
+_ADDRESS_SPACE_LIMIT = 0
+_DEFAULT_ARGUMENTS = [
+    '--exit_on_crash=1',
+    f'--timeout={_TIMEOUT}',
+    f'--fork_server={_SERVER_COUNT}',
+    f'--rss_limit_mb={_RSS_LIMIT}',
+    f'--rlimit_as_mb={_RLIMIT_AS}',
+    f'--address_space_limit_mb={_ADDRESS_SPACE_LIMIT}',
+]
+
+_CRASH_REGEX = re.compile(
+    'centipede.cc:572] ReportCrash[0]: crash detected, saving input to (.*)')
+
+
+class CentipedeError(Exception):
+  """Base exception class."""
+
+
+def _get_runner():
+  """Get the Centipede runner."""
+  centipede_path = os.path.join(environment.get_value('BUILD_DIR'), 'centipede')
+  if not os.path.exists(centipede_path):
+    raise CentipedeError('Centipede not found in build')
+
+  os.chmod(centipede_path, 0o755)
+  if environment.get_value('USE_UNSHARE'):
+    return new_process.UnicodeModifierRunner(centipede_path)
+
+  return new_process.UnicodeProcessRunner(centipede_path)
+
+
+def _get_reproducer_path(line):
+  """Get the reproducer path, if any."""
+  crash_match = _CRASH_REGEX.match(line)
+  if not crash_match:
+    return None
+
+  return crash_match.group(1)
+
+
+class Engine(engine.Engine):
+  """Centipede engine implementation."""
+
+  @property
+  def name(self):
+    return 'centipede'
+
+  def prepare(self, corpus_dir, target_path, build_dir):  # pylint: disable=unused-argument
+    """Prepare for a fuzzing session, by generating options. Returns a
+    FuzzOptions object.
+
+    Args:
+      corpus_dir: The main corpus directory.
+      target_path: Path to the target.
+      build_dir: Path to the build directory.
+
+    Returns:
+      A FuzzOptions object.
+    """
+    os.chmod(target_path, 0o775)
```

Comment:
Do we need this? Maybe we need to do it on the centipede binary as well?

Reply:
Probably not... I noticed [honggfuzz](https://github.com/google/clusterfuzz/blob/388f4cf83daa3aa0695a30de4c0eb446ff828ffd/src/clusterfuzz/_internal/bot/fuzzers/honggfuzz/engine.py#L214) has it, so I did the same just in case.
Removed now.

Reply:
Did you remove this?